presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java

```diff
@@ -0,0 +1,179 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.execution.QueryManager;
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.spi.NodeState;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.inject.Inject;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/")
+public class ClusterStatsResource
+{
+    private final InternalNodeManager nodeManager;
+    private final QueryManager queryManager;
+
+    @Inject
+    public ClusterStatsResource(InternalNodeManager nodeManager, QueryManager queryManager)
+    {
+        this.nodeManager = nodeManager;
+        this.queryManager = queryManager;
+    }
+
+    @GET
+    @Path("/v1/cluster")
+    @Produces(MediaType.APPLICATION_JSON)
+    public ClusterStats getClusterStats()
+    {
+        long runningQueries = 0;
+        long blockedQueries = 0;
+        long queuedQueries = 0;
+
+        long activeNodes = nodeManager.getNodes(NodeState.ACTIVE).size();
+        long runningDrivers = 0;
+        double memoryReservation = 0;
+
+        double rowInputRate = 0;
+        double byteInputRate = 0;
+        double cpuTimeRate = 0;
+
+        for (QueryInfo query : queryManager.getAllQueryInfo()) {
+            if (query.getState().equals(QueryState.QUEUED)) {
+                queuedQueries++;
+            }
+            if (query.getState().equals(QueryState.RUNNING)) {
+                if (query.getQueryStats().isFullyBlocked()) {
+                    blockedQueries++;
+                }
+                else {
+                    runningQueries++;
+                }
+            }
+
+            if (!query.getState().equals(QueryState.FAILED) && !query.getState().equals(QueryState.FINISHED)) {
+                double totalExecutionTimeSeconds = query.getQueryStats().getElapsedTime().toMillis() / 1000.0;
+                if (totalExecutionTimeSeconds != 0) {
+                    byteInputRate += query.getQueryStats().getProcessedInputDataSize().toBytes() / totalExecutionTimeSeconds;
+                    rowInputRate += query.getQueryStats().getProcessedInputPositions() / totalExecutionTimeSeconds;
+                    cpuTimeRate += (query.getQueryStats().getTotalCpuTime().toMillis() / 1000.0) / totalExecutionTimeSeconds;
+                    memoryReservation += query.getQueryStats().getTotalMemoryReservation().toBytes();
```

Comment:
Instead of total "user" memory. It might be more useful to display free distributed memory in the general pool & system pool, along with min local memory on a node for general & system